registry/export.go

```diff
@@ -15,240 +18,200 @@ import (
 	"github.com/go-kivik/kivik"
 )
 
-func Export(out io.Writer) (err error) {
-	buf := bufio.NewWriter(out)
-	defer func() {
-		if err == nil {
-			err = buf.Flush()
-		}
-	}()
-
-	zw := gzip.NewWriter(buf)
-	defer func() {
-		if err == nil {
-			err = zw.Close()
-		}
-	}()
-
-	tw := tar.NewWriter(zw)
-	defer func() {
-		if err == nil {
-			err = tw.Close()
-		}
-	}()
-
-	for _, c := range spaces {
-		if err = writeDocs(c.AppsDB(), tw); err != nil {
-			return
-		}
-		if err = writeDocs(c.VersDB(), tw); err != nil {
-			return
-		}
+const rootPrefix = "registry"
+const couchPrefix = "couchdb"
+const swiftPrefix = "swift"
+const documentSuffix = ".json"
+const contentTypeAttr = "COZY.content-type"
+
+func writeFile(writer *tar.Writer, path string, content []byte, attrs map[string]string) error {
+	header := &tar.Header{
+		Typeflag:   tar.TypeReg,
+		Name:       path,
+		Mode:       0640,
+		Size:       int64(len(content)),
+		PAXRecords: attrs,
 	}
-
-	err = writeDocs(globalEditorsDB, tw)
-	return
+	if err := writer.WriteHeader(header); err != nil {
+		return err
+	}
+	_, err := writer.Write(content)
+	return err
 }
 
-func writeDocs(db *kivik.DB, tw *tar.Writer) error {
-	rows, err := db.AllDocs(ctx, map[string]interface{}{
-		"include_docs": true,
-		"limit":        2000,
-	})
+func writeReaderFile(writer *tar.Writer, path string, reader io.Reader, attrs map[string]string) error {
+	content, err := ioutil.ReadAll(reader)
 	if err != nil {
 		return err
 	}
+	return writeFile(writer, path, content, attrs)
+}
 
-	type attachment struct {
-		docID string
-		name  string
+func exportCouchDocument(writer *tar.Writer, prefix string, db *kivik.DB, rows *kivik.Rows) error {
+	id := rows.ID()
+	if strings.HasPrefix(id, "_design") {
+		return nil
 	}
 
-	dbName := db.Name()
-
-	var atts []*attachment
-	for rows.Next() {
-		if strings.HasPrefix(rows.ID(), "_design") {
-			continue
-		}
+	file := path.Join(prefix, fmt.Sprintf("%s%s", id, documentSuffix))
```

Comment:
How do you guarantee that couchdb document IDs are compatible with filesystem charset requirements ?

Reply:
We never hit the filesystem, only using in-memory tar read/write.
So no problem here I think.

Reply:
ok, understood.